File: 04-stack/00150-Evaluate_Reverse_Polish_Notation/00150-evaluate_reverse_polish_notation.java
// 150: Evaluate Reverse Polish Notation
// https://leetcode.com/problems/evaluate-reverse-polish-notation/


import java.util.Stack;


class Solution {
    // SOLUTION
    public int evalRPN(String[] tokens) {
        Stack<Integer> result = new Stack<>();

        for (var s : tokens) {
            if(s.length()>1 || Character.isDigit(s.charAt(0))) {
                result.push(Integer.parseInt(s));
            } else {
                var x2 = result.peek(); result.pop();
                var x1 = result.peek(); result.pop();
                
                switch(s.charAt(0)) {
                    case '+': x1+=x2; break;
                    case '-': x1-=x2; break;
                    case '*': x1*=x2; break;
                    case '/': x1/=x2; break;
                }

                result.push(x1);
            }
        }

        return result.peek();
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String[] tokens = {"2","1","+","3","*"};

        // OUTPUT
        var result = o.evalRPN(tokens);
        System.out.println(result);
    }
}

File: 04-stack/00150-Evaluate_Reverse_Polish_Notation/00150-evaluate_reverse_polish_notation.go
// 150: Evaluate Reverse Polish Notation
// https://leetcode.com/problems/evaluate-reverse-polish-notation/

package main

import (
	"fmt"
	"strconv"
	"unicode"
)

// SOLUTION
func evalRPN(tokens []string) int {
	var result []int

	for _, s := range tokens {
		if len(s)>1 || unicode.IsDigit(rune(s[0])) {
			si, _ := strconv.ParseInt(string(s), 10, 64)
			result = append(result, int(si))
		} else {
			x2 := result[len(result)-1]; result = result[:len(result)-1]
			x1 := result[len(result)-1]; result = result[:len(result)-1]
			
			switch s[0] {
				case '+': x1+=x2; break
				case '-': x1-=x2; break
				case '*': x1*=x2; break
				case '/': x1/=x2; break
			}

			result = append(result, x1)
		}
	}

	return result[len(result)-1]
}

func main() {
	// INPUT
	tokens := []string{"2","1","+","3","*"}

	// OUTPUT
	result := evalRPN(tokens)
	fmt.Println(result)
}